src/shogun/regression/LeastAngleRegression.cpp

```diff
@@ -75,96 +58,131 @@ static void find_max_abs(const vector<float64_t> &vec, const vector<bool> &ignor
 	}
 }
 
-CLeastAngleRegression::CLeastAngleRegression(bool lasso) :
-	CLinearMachine(), m_lasso(lasso),
-	m_max_nonz(0), m_max_l1_norm(0)
+template <typename ST>
+void CLeastAngleRegression::plane_rot(ST x0, ST x1,
+	ST &y0, ST &y1, SGMatrix<ST> &G)
 {
-	m_epsilon = CMath::MACHINE_EPSILON;
-	SG_ADD(&m_epsilon, "epsilon", "Epsilon for early stopping", MS_AVAILABLE);
-	SG_ADD(&m_max_nonz, "max_nonz", "Max number of non-zero variables", MS_AVAILABLE);
-	SG_ADD(&m_max_l1_norm, "max_l1_norm", "Max l1-norm of estimator", MS_AVAILABLE);
-}
+	memset(G.matrix, 0, G.num_rows * G.num_cols * sizeof(ST));
 
-CLeastAngleRegression::~CLeastAngleRegression()
-{
+	if (x1 == 0)
+	{
+		G(0, 0) = G(1, 1) = 1;
+		y0 = x0;
+		y1 = x1;
+	}
+	else
+	{
+		ST r = CMath::sqrt(x0*x0+x1*x1) ;
+		ST sx0 = x0 / r;
+		ST sx1 = x1 / r;
 
+		G(0,0) = sx0;
+		G(1,0) = -sx1;
+		G(0,1) = sx1;
+		G(1,1) = sx0;
+
+		y0 = r;
+		y1 = 0;
+	}
 }
 
-bool CLeastAngleRegression::train_machine(CFeatures* data)
+bool CLeastAngleRegression::train_machine(CFeatures * data)
 {
+	//error checking
 	if (!m_labels)
 		SG_ERROR("No labels set\n")
+
 	if (m_labels->get_label_type() != LT_REGRESSION)
-		SG_ERROR("Expected RegressionLabels\n")
+		SG_ERROR("Provided labels (%s) are of type (%d), where they shold be regressionlabels (%d)\n",
+			m_labels->get_name(), m_labels->get_label_type(), LT_REGRESSION, m_labels->get_label_type())
 
 	if (!data)
-		data=features;
+	{
+		if(!features)
+			SG_ERROR("Both data and LinearMachine's features is null! Please ensure that one of them is set\n")
 
-	if (!data)
-		SG_ERROR("No features set\n")
+		SG_INFO("Data is null, so using LinearMachine's features instead.\n")
+
+		if(features->get_feature_class() != C_DENSE)
+			SG_ERROR("LinearMachine's features' feature-type is not C_DENSE. Please ensure that it is\n")
+			
+		data = features;
+	}
+	else
+		if (data->get_feature_class() != C_DENSE)
+			SG_ERROR("Expected Simple Features (C_DENSE)")
+
+	if (data->get_num_vectors() != m_labels->get_num_labels())
+		SG_ERROR("Number of training vectors (%d) does not match number of labels (%d)\n", 
+			data->get_num_vectors(), m_labels->get_num_labels())
 
-	if (m_labels->get_num_labels() != data->get_num_vectors())
-		SG_ERROR("Number of training vectors does not match number of labels\n")
+	if (data->get_feature_type() != F_DREAL && data->get_feature_type() != F_SHORTREAL)
+		SG_ERROR("Expected Real Features!\n")
 
-	if (data->get_feature_class() != C_DENSE)
-		SG_ERROR("Expected Simple Features\n")
+	//check for type of CFeatures, then call the appropriate template method
+	if(data->get_feature_type() == F_DREAL)
+		return CLeastAngleRegression::train_machine_i((CDenseFeatures<float64_t> *) data);
+	else//data->get_feature_type() == F_DREAL
```

Comment:
@vigsterkr 
Can't we somehow do these type of checks in a base class? This will appear over and over again if we start porting more algos to using multiple memory widths